TheJumpCloud/support
# NB-12742: Added PWM MacOS architecture detection, redirection to ARM download

## Issues
* [NB-12742](https://jumpcloud.atlassian.net/browse/NB-12742) - Update PWM Installation script to support ARM builds

## Dependencies
- https://github.com/jc-pny/desktop-app-v2/pull/989

## What does this solve?
It adds the ability for the script to download arm64 PWM builds directly instead of Intel based builds.

## Is there anything particularly tricky?
No, just to note that when the [dependency PR](https://github.com/jc-pny/desktop-app-v2/pull/989) gets merged, we will be able to successfully test it on the prod cdn. (using the prod link)
This is how the prod link for MacOS ARM builds will look like:
https://cdn.pwm.jumpcloud.com/DA/release/arm64/JumpCloud-Password-Manager-latest.dmg
Staging link (just for reference):
https://cdn.pwm.stg01.jumpcloud.com/DA/staging/release/arm64/JumpCloud-Password-Manager-Staging-latest.dmg

## How should this be tested?
Testing is done only on MacOS, no need for any other OS.
***NOTE: If you're to test on staging (before the release is merged & deployed) please check the end of the `How should this be tested?` block right after step 2.

1. Make sure your device is `arm64` architecture (you can run `uname -m` in your terminal, output should be `arm64`)
2.  Create a `.sh` file containing the following:

```
#!/bin/bash

# This script will install password manager in Users/$user/Applications for all user accounts based on their architecture (x64 or arm64)
DownloadUrl="https://cdn.pwm.jumpcloud.com/DA/release/JumpCloud-Password-Manager-latest.dmg"

# Detect device architecture
DeviceArchitecture=$(uname -m)


if [ "$DeviceArchitecture" = "arm64" ]; then
    DownloadUrl="https://cdn.pwm.jumpcloud.com/DA/release/arm64/JumpCloud-Password-Manager-latest.dmg"
fi


regex='^https.*.dmg$'
if [[ $DownloadUrl =~ $regex ]]; then
    echo "URL points to direct DMG download"
    validLink="True"
else
    echo "Searching headers for download links"
    urlHead=$(curl -s --head $DownloadUrl)

    locationSearch=$(echo "$urlHead" | grep https:)

    if [ -n "$locationSearch" ]; then

        locationRaw=$(echo "$locationSearch" | cut -d' ' -f2)

        locationFormatted="$(echo "${locationRaw}" | tr -d '[:space:]')"

        regex='^https.*'
        if [[ $locationFormatted =~ $regex ]]; then
            echo "Download link found"
            DownloadUrl=$(echo "$locationFormatted")
        else
            echo "No https location download link found in headers"
            exit 1
        fi

    else

        echo "No location download link found in headers"
        exit 1
    fi

fi



#Create Temp Folder
DATE=$(date '+%Y-%m-%d-%H-%M-%S')

TempFolder="Download-$DATE"

mkdir /tmp/$TempFolder

# Navigate to Temp Folder
cd /tmp/$TempFolder

# Download File into Temp Folder
curl -s -O "$DownloadUrl"

# Capture name of Download File
DownloadFile="$(ls)"

echo "Downloaded $DownloadFile to /tmp/$TempFolder"

# Verifies DMG File
regex='\.dmg$'
if [[ $DownloadFile =~ $regex ]]; then
    DMGFile="$(echo "$DownloadFile")"
    echo "DMG File Found: $DMGFile"
else
    echo "File: $DownloadFile is not a DMG"
    rm -r /tmp/$TempFolder
    echo "Deleted /tmp/$TempFolder"
    exit 1
fi

# Mount DMG File -nobrowse prevents the volume from popping up in Finder

hdiutilAttach=$(hdiutil attach /tmp/$TempFolder/$DMGFile -nobrowse)

echo "Used hdiutil to mount $DMGFile "

err=$?
if [ ${err} -ne 0 ]; then
    echo "Could not mount $DMGFile Error: ${err}"
    rm -r /tmp/$TempFolder
    echo "Deleted /tmp/$TempFolder"
    exit 1
fi

regex='\/Volumes\/.*'
if [[ $hdiutilAttach =~ $regex ]]; then
    DMGVolume="${BASH_REMATCH[@]}"
    echo "Located DMG Volume: $DMGVolume"
else
    echo "DMG Volume not found"
    rm -r /tmp/$TempFolder
    echo "Deleted /tmp/$TempFolder"
    exit 1
fi

# Identify the mount point for the DMG file
DMGMountPoint="$(hdiutil info | grep "$DMGVolume" | awk '{ print $1 }')"

echo "Located DMG Mount Point: $DMGMountPoint"

# Capture name of App file

cd "$DMGVolume"

AppName="$(ls | Grep .app)"

cd ~

echo "Located App: $AppName"


DMGAppPath=$(find "$DMGVolume" -name "*.app" -depth 1)

userInstall=false

for user in $(dscl . list /Users | grep -vE 'root|daemon|nobody|^_')
do
    if [[ -d /Users/$user ]]; then
        # Create ~/Applications folder
        if [[ ! -d /Users/$user/Applications ]]; then
            mkdir /Users/$user/Applications
        fi
        if [[ -d /Users/$user/Applications/JumpCloud\ Password\ Manager.app ]]; then
            # remove if exists
            rm -rf /Users/$user/Applications/JumpCloud\ Password\ Manager.app
        fi

        # Copy the contents of the DMG file to /Users/$user/Applications/
        # Preserves all file attributes and ACLs
        cp -pPR "$DMGAppPath" /Users/$user/Applications/

        # Change ownership of the file to the user of this loop iteration
        chown -v $user /Users/$user/Applications/JumpCloud\ Password\ Manager.app

        if [[ -d /Users/$user/Desktop/JumpCloud\ Password\ Manager.app ]]; then
            # remove alias on desktop if exists
            rm -rf /Users/$user/Desktop/JumpCloud\ Password\ Manager.app
        fi

        err=$?
        if [ ${err} -ne 0 ]; then
            echo "Could not copy $DMGAppPath Error: ${err}"
            hdiutil detach $DMGMountPoint
            echo "Used hdiutil to detach $DMGFile from $DMGMountPoint"
            rm -r /tmp/$TempFolder
            echo "Deleted /tmp/$TempFolder"
            exit 1
        fi

        userInstall=true
        echo "Copied $DMGAppPath to /Users/$user/Applications"

        # Create an alias on desktop
        ln -s /Users/$user/Applications/JumpCloud\ Password\ Manager.app /Users/$user/Desktop/JumpCloud\ Password\ Manager.app
    fi
done


# Check if password manager is installed in /Applications; remove if we installed in user directory
if [ -d /Applications/JumpCloud\ Password\ Manager.app ] && [ $userInstall = true ]; then
    # remove if exists
    rm -rf /Applications/JumpCloud\ Password\ Manager.app
fi

# Unmount the DMG file
hdiutil detach $DMGMountPoint

echo "Used hdiutil to detach $DMGFile from $DMGMountPoint"

err=$?
if [ ${err} -ne 0 ]; then
    abort "Could not detach DMG: $DMGMountPoint Error: ${err}"
fi

# Remove Temp Folder and download
rm -r /tmp/$TempFolder

echo "Deleted /tmp/$TempFolder"

exit
```
3. IF you're testing with the staging link, after waiting for the app to download you'll notice a blank shortcut on your desktop. Ignore this as the script manually creates this based on the prod app name (theyre different, its a blank shortcut, this wont happen using the prod link)
4. Go to `~/Applications/`
5. You should see `JumpCloud Password Manager` if you're using the prod link, `JC Password Manager Staging` if you're using staging
6. Right click on the app -> Get info -> You should see `Apple Silicon` as the app type (Screenshot attached)


If you're testing on staging, you should modify the following in the `.sh` file:
```
if [ "$DeviceArchitecture" = "arm64" ]; then
    DownloadUrl="https://cdn.pwm.stg01.jumpcloud.com/DA/staging/release/arm64/JumpCloud-Password-Manager-Staging-latest.dmg"
fi
```
## Screenshots
![image](https://github.com/user-attachments/assets/1860ce26-f22a-4309-94af-cb3c8b749ae3)


[NB-12742]: https://jumpcloud.atlassian.net/browse/NB-12742?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
